### ISSUE TITLE
Native-like desktop app

### ISSUE THREAD

**Issue by ayyaruq** [Supamiu/ffxiv-teamcraft#502]
The desktop app feels like a webpage running standalone at the moment, particularly on mac where it has a bunch of features that just don't work right for the OS at all. This issue is to track the broken components as well as features to make it feel more like a native app, and assist development and debugging.

If there are any non-mac issues that are raised here, please leave a comment so we can track appropriately. For individual fixes or features it might be worth having a proper github project. Also, someone to test on Linux.

This is all the major stuff I can think of but I'm sure there's more. TL;DR app should feel native on the platform instead of a wrapper, and not have weird windowing bugs.

#### Bugs (mac):
- [ ] closing the window reloads app and adds a new icon to menubar (so you get multiple icons)
- [ ] app is constantly in menubar plus dock, and highlighted even when not targeted in menubar
- [ ] settings can't be opened by global preferences shortcut
- [ ] custom windowing chrome for windows breaks window framing
- [ ] missing app icon
- [ ] language in app header unnecessary - use system preference or settings override
- [ ] version in app header unnecessary - use "about teamcraft" in menu
- [ ] nav has scrollbar from unnecessary elements (about, settings, maybe user account block)

#### Features (mac/other?):
- [ ] native windowing - just use the OSX window chrome, even if the wrapper is custom (like discord)
- [ ] back/forward buttons probably aren't needed in the app (discussion welcome)
- [ ] Open Teamcraft Link needs a better home
- [ ] Timers button in bottom right needs a better home
- [ ] any pages with their own settings popup should probably move to a tab in settings page (see discord/slack)
- [ ] settings should be a full-view modal page, overlaying the current page but not replacing it (see discord/slack)
- [ ] menubar icon should have quit/preferences/alarms/mute/quick-macro-translate/update
- [ ] setting to exit app when closing, vs hide from dock and keep in menubar

**Comment by Sanaki** [comment]
Should this come to fruition, I could test builds on Linux.

**Comment by cking** [comment]
any updates?
:3c

**Comment by stone-solid** [comment]
It's been 2 years since this has seen any love. Any chances this is being worked on? I play FFXIV on Mac, and would absolutely love to be able to use all FFXIV Teamcraft features.

**Comment by Supamiu** [comment]
You can still, using web browser, but as I don't have a Mac to work on, it's frozen until someone starts working on it.

**Comment by SapphicFire** [comment]
Commenting to remind myself later, would like to offer my hand and start looking at this.

**Comment by ayyaruq** [comment]
@SapphicFire I've straight up not had any time to work on this but if you swing by our discord I'm more than happy to answer any questions, run by how I'd structure this, or just point you in the right direction if you need it. I think there's some more general changes we need to make to how the app initialises which will make all this much easier but we need to make a start on that at some point anyway.

**Comment by stone-solid** [comment]
i'm incredibly very new to coding but i'm really comfortable with designing. as i learn coding, i'll be seeing how i can help too. if there's anything i can do design-wise, i'll be glad to help with that as well.

**Comment by ayyaruq** [comment]
I finally got a new laptop where I have enough RAM to actually address this. Will get some stuff done this weekend. Have updated the original post body with bugs/features that have already been addressed, and some new ones that I came up with.

**Comment by cking** [comment]
> native windowing - custom titlebar (like discord, not necessary for linux/win), nice to have, not required
#531
not sure if it is still implemented but i dont think it got removed

**Comment by ayyaruq** [comment]
It's no longer a thing, although the code for handling it is still there. `options.nativeDecorator` seems to never be set, but either way doesn't have any effect on macos.

The nicest way seems to be `titleBarStyle: 'hidden'` in the BrowserWindow creation to get some extra real estate, but it runs into the issue of the app not having any way to drag around the screen, and zooming smaller makes it like 5x1 pixels for some reason.

**Comment by cking** [comment]
well i prefer native titlebars anyway, be it on windows, or now for me on linux (dont own a windows machine anymore)

**Comment by Xeftax** [comment]
# Packet capture working on macOS with the official Square Enix client (CrossOver/Wine 7.7)

Got live packet capture working on macOS against the **official SE client**
(`/Applications/FINAL FANTASY XIV ONLINE.app`, which is a bundled CrossOver
engine reporting Wine 7.7) — no XIVLauncher, no Proton, and the game keeps
running on its own stock engine. Writing up the failure modes because every one
of them produces a misleading error.

Environment: Apple Silicon, macOS 15, Teamcraft 11.4.30 built from source,
official SE client.

## Step 0: building the app, since there's no macOS release

The releases page only ships Windows and Linux artifacts, so the app has to be
built locally. The build works out of the box on Apple Silicon, but it needs a
toolchain a non-developer won't have yet. From a completely clean Mac:

**Prerequisites** (no Homebrew needed, ~6 GB of free disk space). These mirror
`CONTRIBUTING.md`, which is worth reading first:

1. **Xcode Command Line Tools** — provides `git` and the compilers:
   ```bash
   xcode-select --install
   ```
   A dialog appears; accept and wait. Full Xcode is *not* required.

2. **Node.js 22.14 or newer** — the version `CONTRIBUTING.md` requires.
   Download the macOS `.pkg` installer from <https://nodejs.org> and run it;
   it also installs `npm` and `npx`, so there's nothing else to do for those.
   This build was done on Node 25, where the only symptom is the `re2` warning
   below.

3. **Yarn** — the repo uses Yarn Classic (a `yarn.lock` v1 file):
   ```bash
   npm install -g yarn
   ```

Check they're all there before starting — each should print a version:

```bash
git --version && node --version && yarn --version && npx --version
```

**Build:**

```bash
git clone https://github.com/ffxiv-teamcraft/ffxiv-teamcraft.git
cd ffxiv-teamcraft
yarn install
yarn prebuild
yarn electron:build
yarn build:electron-prod
npx electron-builder --mac --arm64
```

Output lands in `release/mac-arm64/FFXIV Teamcraft.app`. Expect the clone plus
`node_modules` to weigh around 5-6 GB, and the whole thing to take a while —
`yarn install` and the Angular production build are both long.

During `yarn install` the native module `re2` may fail to compile. Ignore it:
it's an optional dependency, `yarn install` carries on, and neither the build
nor the app cares.

### Gatekeeper flags the first build as malware

On first launch macOS may not just block the app but declare it malware and
move it straight to the trash. `spctl -a -vv` reports `revoked`.

It is **not** the bundled Windows binaries (`deucalion-bridge.exe`,
`deucalion*.dll`) — that was my first assumption and it's wrong; they can stay
in the bundle. It's the generic ad-hoc signature `electron-builder` produces by
default. Re-signing the bundle yourself fixes it:

```bash
codesign --force --deep --sign - "FFXIV Teamcraft.app"
xattr -cr "/Applications/FFXIV Teamcraft.app"
```

After that `spctl` reports `rejected` instead of `revoked` — i.e. the normal
unsigned-app block, which right-click → Open gets past. Then copy the bundle to
`/Applications`.

## The three blockers

### 1. The bridge must run under the *game's own* Wine engine

This is the big one, and it's what makes the app look like it "crashes FFXIV".

`deucalion-bridge.exe` has to inject into a process owned by the game's
`wineserver`. If `wineBin` points at *any other* Wine build — even the same
version number — you get two incompatible `wineserver` instances on the same
`WINEPREFIX`:

```
wine client error:0: version mismatch 755/930.
```

(755 = the SE engine's protocol, 930 = WineHQ 11.0.) Whichever process starts
second kills the other's session, in both directions: enabling capture kills a
running game, and launching the game kills a running bridge. It also rewrites
`system.reg`/`user.reg` in the bottle each time, which eventually corrupts the
prefix badly enough that the game won't start even after you restore the
original binaries.

So on macOS, `wineBin` must be the game's own loader:

```
/Applications/FINAL FANTASY XIV ONLINE.app/Contents/SharedSupport/finalfantasyxiv/FINAL FANTASY XIV ONLINE/wineloader64
```

That binary needs `WINELOADER`/`WINESERVER`/`WINEDLLPATH`/
`DYLD_FALLBACK_LIBRARY_PATH` set, and the Wine binary field can't carry them
(the "extra env" field splits on whitespace and these paths contain spaces),
hence a wrapper script — see below.

Note the CrossOver `wine` perl wrapper in the same directory exits silently
without doing anything and prints nothing, even with `WINEDEBUG` set. Use
`wineloader64` directly or you will debug nothing.

### 2. esync must match

The SE client's `wineserver` runs with esync. A client without it is refused:

```
err:esync:esync_init Server is running with WINEESYNC but this process is not,
please enable WINEESYNC or restart wineserver.
```

`WINEESYNC=1` (two E's — not `WINESYNC`) in the extra-env field, or in the
wrapper.

### 3. `bcryptprimitives.dll`

`deucalion-bridge.exe` imports `ProcessPrng` from `bcryptprimitives.dll`. Wine
only gained that stub in mainline around July 2023 (MR !2228), so the SE
engine's 7.7 doesn't have it.

Dropping the real Microsoft DLL from a Windows ISO into the bottle does *not*
work — it loads but fails to initialize under 7.7:

```
err:module:LdrInitializeThunk "bcryptprimitives.dll" failed to initialize, aborting, status c0000142
```

What does work: take the **builtin** `bcryptprimitives.dll` that any Wine 8+
build ships. It's a few KB, exports `ProcessPrng`, and imports nothing but
`advapi32!SystemFunction036` and `kernel32` — so it runs fine under 7.7.

Easiest source, no Apple Developer account and no Heroic needed: install the
regular **Wine Stable** app for macOS (<https://github.com/Gcenx/macOS_Wine_builds>
/ the winehq macOS download), then lift one file out of it:

```
/Applications/Wine Stable.app/Contents/Resources/wine/lib/wine/x86_64-windows/bcryptprimitives.dll
```

⚠️ You are harvesting **one 5 KB file** from that install. Do **not** then point
Teamcraft's "Wine binary" setting at Wine Stable — that's blocker #1 above and
it will kill your game session. Wine Stable is only a convenient box the file
comes in; you can delete it afterwards.

(Apple's Game Porting Toolkit 3.0-3 has an equivalent copy at
`Contents/Resources/wine/lib/wine/x86_64-windows/bcryptprimitives.dll`, 40 KB,
same two imports. Both were tested and both work.)

Copy it into `<bottle>/drive_c/windows/system32/` — that's the location that
actually gets picked up. Dropping it in the SE engine's own
`lib/wine/x86_64-windows/` instead is *not* enough: 7.7 doesn't know the name
as a builtin and still reports

```
err:module:import_dll Library bcryptprimitives.dll (which is needed by
L"...deucalion-bridge.exe") not found
```

And delete the Microsoft one if you already put it there.

No engine swap is needed — one small file, and the game keeps running on its
stock SE engine.

## One app-side bug found along the way

The tray "Packet Capture" checkbox desyncs and then kills a working
capture. `tray-menu.ts` builds the menu once at startup and snapshots
`checked` from `store.get('machina')` at that moment. Nothing ever re-syncs it.
Enable capture from Settings and the tray item still thinks it's off — clicking
it calls `startPcap()` on top of a running capture, the second bridge can't
bind 31594, and the whole thing dies until you quit and relaunch the app.

Toggling from the store instead of from `menuItem.checked` fixes it:

```ts
click: (menuItem) => {
  const enabled = !this.store.get<boolean>('machina', false);
  this.store.set('machina', enabled);
  menuItem.checked = enabled;
  this.mainWindow.win?.webContents.send('toggle-pcap:value', enabled);
  enabled ? this.pcap.startPcap() : this.pcap.stop();
}
```

Opened as #3305.

## Working setup, for anyone else on macOS

0. Build and re-sign the app (see "Step 0" above), install it to
   `/Applications`.

1. Copy `bcryptprimitives.dll` out of any Wine 8+ build (see blocker 3) into
   `~/Library/Application Support/FINAL FANTASY XIV ONLINE/Bottles/published_Final_Fantasy/drive_c/windows/system32/`
   — and do not point Teamcraft at that Wine.

2. Save this as `~/ffxiv-bridge-wine.sh`, `chmod +x` it:

```sh
#!/bin/sh
# Runs the Teamcraft packet-capture bridge under the game's own Wine engine.
#
# Any other Wine build on the same WINEPREFIX gives "wine client error:
# version mismatch" and kills the running game session, in both directions.
# WINEESYNC=1 is required too: the game's wineserver runs esync and refuses
# clients without it. The paths contain spaces, which is why this has to be a
# wrapper rather than Teamcraft's "extra env" field (it splits on whitespace).
SS="/Applications/FINAL FANTASY XIV ONLINE.app/Contents/SharedSupport/finalfantasyxiv"
ENG="$SS/FINAL FANTASY XIV ONLINE"
export WINEESYNC=1
export WINELOADER="$ENG/wineloader64"
export WINESERVER="$ENG/wineserver"
export WINEDLLPATH="$SS/lib/wine"
export DYLD_FALLBACK_LIBRARY_PATH="$SS/lib"
exec "$ENG/wineloader64" "$@"
```

3. Teamcraft → Settings → Packet capture:
   - Wine/Proton prefix: `~/Library/Application Support/FINAL FANTASY XIV ONLINE/Bottles/published_Final_Fantasy`
   - Wine binary: `~/ffxiv-bridge-wine.sh`
   - Extra env: anything (the wrapper sets what matters)

4. Start the game first, then enable packet capture from **Settings** (not
   the tray, until #3305 lands).

Result:

```
[bridge] deucalion.dll injected
[bridge] Listening on 127.0.0.1:31594
DEUCALION: SERVER HELLO. VERSION: 1.5.0. HOOK STATUS: RECV ON. SEND ON. SEND_LOBBY ON. CREATE_TARGET ON.
```

Not just the handshake — live data actually flows: switching jobs in game is
picked up and reflected in Teamcraft straight away.

Game keeps running, no prefix corruption.

## Warning for anyone experimenting

Back up `system.reg` and `user.reg` from the bottle before running *any*
different Wine build against it, even once. Mixing engines rewrites the
registry in incompatible ways and the damage survives restoring the original
binaries — it took a registry restore from backup to get the game launching
again.
